src/qs_codec/encode.py

```
@@ -446,14 +366,273 @@ def _encode(
+            max_depth=_max_depth,
         )
+    ]
+
+    while stack:
+        frame = stack[-1]
+
+        if frame.phase == "start":
+            if frame.max_depth is None:
+                frame.max_depth = _get_max_encode_depth(None)
+            if frame.depth > frame.max_depth:
+                raise ValueError("Maximum encoding depth exceeded")
```

Ruff (0.15.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)